crates/planner/src/builder.rs

```diff
@@ -600,48 +691,327 @@ impl<'a> Context<'a> {
         path.pop();
     }
 
-    fn add_fetch_entity(
+    // Method to handle fields with the @requires directive
+    fn handle_requires_directive(
         &mut self,
-        path: &ResponsePath<'a>,
+        field: &'a Field,
+        field_definition: &'a MetaField,
+        parent_type: &'a MetaType,
+        current_service: &'a str,
+        requires: &'a KeyFields,
         selection_ref_set: &mut SelectionRefSet<'a>,
         fetch_entity_group: &mut FetchEntityGroup<'a>,
+        path: &mut ResponsePath<'a>,
+    ) {
+        // Check if the path length exceeds a reasonable limit to prevent infinite loops
+        if path.len() > 10 {
```

Comment:
Very nice, please - extract `10` into a constant with good name so it will be possible to identify better.